File: src/junction.py
##
# @file: junction.py
# @author: Chris Blatchley
# @author: Thad Bond
# @date: Sun, Feb 23, 2014
# @version: 0.2
# ##
# Junction object implementation.
##

class Junction(object):
    """Junction object to control behavior in edge junctions"""

    ##
    # Junction Constructor
    # @param id		The string identification
    # @param shape	The shape of the junction
    # @return 	Initialized Junction Object
    ##
    def __init__(self, uid, shape):
        super(Junction, self).__init__()
        self.uid = uid
        self.shape = shape
        self.queue = [] #The vehicles waiting to pass through

    ##
    # runTimestep
    ##
    def runTimestep(self):
        if self.shape is "allstop":
            if self.queue: #We have at least something in the queue
                vehicle = self.queue[0]
                newEdge = vehicle.route.getNextEdge(vehicle.currEdge) #Get the next edge for the vehicle
                print("Junction ", self.uid, " is passing a vehicle to edge ", newEdge.uid)
                newEdge.addVehicle(vehicle) #Add the vehicle to the new edge
                vehicle.currSpeed = 0 #Make the vehicle start at 0 speed from stop
                self.queue.remove(vehicle) #The vehicle has passed through, get rid of it

    ##
    # addToQueue
    # @param vehicle The vehicle to add to the queue
    ##
    def addToQueue(self, vehicle):
        #Add it to our queue list
        self.queue.append(vehicle)
        #Remove the vehicle from its previous edge now that its done there
        vehicle.currEdge.removeVehicle(vehicle)

File: src/junction.cpp
#include "junction.h"
using namespace std;

/**
Constructor for Junction object
@param uid		The name or unique identifier of this junction
@param shape	The style of junction logic this junction performs
*/
Junction::Junction( string uid, Shape shape )
{
	_uid = uid;
	_shape = shape;
}

/**
Perform the junction logic for a timestep
*/
void Junction::runTimestep()
{
	switch ( _shape )
	{
		case Shape::Throughway:
			while ( !_waitQueue.empty() )
			{
				_waitQueue.front(); //TODO: SEND TO NEXT EDGE ON ROUTE
				_waitQueue.pop();
			}
			break;
		case Shape::AllStop:
			_waitQueue.front(); //TODO: SEND TO NEXT EDGE ON ROUTE
			_waitQueue.pop();
			break;
	}
}

/**
Add a vehicle to the junction's queue
@param vehicle	The vehicle to add
*/
void Junction::queueVehicle( Vehicle* vehicle )
{
	//Add the vehicle to the wait queue
	_waitQueue.push( vehicle );
}

/**
Destructor for the junction object
*/
Junction::~Junction( void )
{
}
